staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,278 @@
+// StorageVersion for CRDs. It does goroutine management to allow CRD
+// storage version updates running in the background and not blocking the caller.
+type Manager struct {
+	// client is the client interface that manager uses to update
+	// StorageVersion objects.
+	client genericstorageversion.Client
+	// apiserverID is the ID of the apiserver that invokes this manager.
+	apiserverID string
+	crdInformer crdinformers.CustomResourceDefinitionInformer
+	svInformer  svInformers.StorageVersionInformer
+	// workqueue to process storageversion updates for this CRD.
+	queue *workqueue.Type
+	// map from crd.Name -> number of active teardowns for this CRD.
+	teardownCountMap sync.Map
+}
+
+// NewManager creates a CRD StorageVersion Manager.
+func NewManager(svClient genericstorageversion.Client, apiserverID string,
+	crdInformer crdinformers.CustomResourceDefinitionInformer, svInformer svInformers.StorageVersionInformer) *Manager {
+	return &Manager{
+		client:           svClient,
+		apiserverID:      apiserverID,
+		crdInformer:      crdInformer,
+		svInformer:       svInformer,
+		queue:            workqueue.NewNamed("storageversion-updater"),
+		teardownCountMap: sync.Map{},
+	}
+}
+
+// UpdateActiveTeardownsCount updates the teardown count of the CRD in sync.Map.
+func (m *Manager) UpdateActiveTeardownsCount(crdName string, value int) error {
+	val, ok := m.teardownCountMap.Load(crdName)
+	if !ok {
+		return fmt.Errorf("error while trying to update number of teardowns. No entry found in sync.Map for %v", crdName)
+	}
+	newTeardownCount := val.(int) + value
+	m.teardownCountMap.Store(crdName, newTeardownCount)
+	return nil
+}
+
+// Enqueue adds the CRD name to the SV upadte queue.
+func (m *Manager) Enqueue(crdName string) {
+	if crdName == "" {
+		return
+	}
+	klog.V(4).Infof("enqueueing crd %s for storageversion update", crdName)
+	m.queue.Add(crdName)
+}
+
+// DeleteTeardownCountInfo deletes specified key from the sync.Map.
+func (m *Manager) DeleteTeardownCountInfo(crdName string) {
+	m.teardownCountMap.Delete(crdName)
+}
+
+// Sync runs a goroutine over the SV updater queue to indefinitely
+// process any queued storageversion updates unless stopCh is invoked.
+func (m *Manager) Sync(stopCh <-chan struct{}, workers int) {
+	defer m.queue.ShutDownWithDrain()
+	klog.V(2).Infof("starting storageversion sync loop")
+	for i := 0; i < workers; i++ {
+		go wait.Until(func() {
+			m.worker()
+		}, workerFrequency, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (m *Manager) worker() {
+	for m.processLatestUpdateFor() {
+	}
+}
+
+func (m *Manager) processLatestUpdateFor() bool {
+	ctx := context.TODO()
+	key, quit := m.queue.Get()
+	defer m.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	// Fetch the latest CRD from CRD informer cache to update the storageversion for it.
+	crdName := key.(string)
+	crd, err := m.crdInformer.Lister().Get(crdName)
+	if err != nil {
+		klog.V(4).Infof("crd: %s not found in CRD cache, abandoning storageversion update", crdName)
```

Comment:
Will there be errors other than NotFound? Can we distinguish them? If the CRD doesn't exists, we don't need to write SV

Reply:
Added separate handling for NotFound err - i.e. do not requeue SV update when this happens